add blogpost on comment preserving cabal parser (#27)
haskell/blog.haskell.org: content/cabal-comment-preserving-parser/index.md

+++
title = "Comment Preserving Cabal Parser"
date = 2025-12-25
[taxonomies]
authors = ["Léana Jiang"]
categories = ["Cabal"]
# tags = ["Practices"]
+++

# Motivation
Cabal [^1] is the standard package system for Haskell software, it is similar to Cargo from the rust
world, or OPAM if you come from OCaml.

Cabal reads cabal package manifests in the cabal format (with the extension `.cabal`).
However, it is currently unable to modify it loselessly: comments are lost, imports are
fused and written in place, `elif` in a conditional will be desugared to a nested `if` in an `else`.

> For this reason, criticial features such as adding modules to the .cabal manifest or generating

Reviewer: ```suggestion
For this reason, critical features such as adding modules to the .cabal manifest or generating
```
Or was it meant to be "crucial"?

Author: Thanks, nice catch :+1:

+++

_This article was originally posted on my blog [here](https://confusedcompiler.org/articles/2025-12-comment-preserving-cabal-parser)._
<hr/>

# Motivation
Cabal [^1] is the standard package system for Haskell software, it is similar to Cargo from the rust
world, or OPAM if you come from OCaml.

Cabal reads cabal package manifests in the cabal format (with the extension `.cabal`).
However, it is currently unable to modify it loselessly: comments are lost, imports are
fused and written in place, `elif` in a conditional will be desugared to a nested `if` in an `else`.

For this reason, critical features such as adding modules to the .cabal manifest or generating
package dependency bounds are implemented poorly.\
For example, cabal emits a warning when a module is not listed in the manifest, but it can't apply
the fix the warning indicates for you.\
Furthermore, `cabal gen-bounds` (which generates dependency bounds that are not specified) doesn't
write dependency bounds information directly to the cabal file, it dumps them to the terminal.
Frustrating!

If we use the current pretty printer in cabal to try to implement these features, the resulting
cabal manifest will be mangled.

> This is because cabal can't do _exact printing_ yet.

Reviewer 2: ```suggestion
This is because cabal can't do _exact-print parsing_ yet.
```

Author: I think this is correct? By "exact-print" I wanted to say it can't do exact print nor exact parse.

Reviewer 2: @leana8959 ok. :)

Concretely this means the spaces and empty lines, comments, common stanzas, and if conditions that
you add will be lost.

> Adding an exact printer will allow cabal to modify the cabal files automatically and loselessly!

Reviewer 2: ```suggestion
Adding an exact-print parser will allow `cabal` to parse a package manifest and modify it without losing any details (_losslessly_)!
```

Author: Here too, I meant to say modify in the sense of modifying a cabal file on disk, so it would require both exact-parse and exact-print.

An exact printer is a special kind of pretty printer.
It uses concrete syntax information in the in-memory representation to output a file that is
byte-to-byte identical to the file originally parsed.
To do that, its corresponding parser must store enough information in the in-memory
representation.
Formally, the exact parser/printer combination obeys the law presented below.
```hs
forall cabalFile.
  IsValid cabalFile =>
    exactPrint (exactParse cabalFile) == cabalFile
```

> This reads as "forall valid cabal manifest, the operation `exactPrint . exactParse` should output a

Reviewer 2: Either say "for all valid manifests" (plural) or "for any valid cabal manifest" (singular). :)

Author: I chose the former :+1:

file that is byte-to-byte identical to the original".

Furthermore, we also aim to allow modification to the in-memory representation. Unmodified parts of
the representation will be printed out verbatim to the original parsed cabal file.

This property is crucial for tools to be able to manipulate cabal files programmatically:
adding dependency, export a module, or generate package bounds just to name a few.

> A tracking issue was opened in cabal's repository since 2021 [here](https://github.com/haskell/cabal/issues/7544).

Reviewer 2: This sentence feels pretty alone amongst all the explanations. It should be put in the context section at the beginning, to inform the reader about the project in itself.

Author: The linebreak before this line indeed makes it out of place. Would moving it to right after the line describing the law make it better?

> This reads as "for all valid cabal manifests, the operation `exactPrint . exactParse` should output a file that is byte-to-byte identical to the original". This property is crucial for tools to be able to manipulate cabal files programmatically: adding dependency, export a module, or generate package bounds just to name a few.

Reviewer 2: Sorry, GitHub is confusing here, I specifically meant line 55. :)

Author: I moved line 55 upwards :+1:

> The first step to achieve an exact printer is to preserve concrete syntax information. For example,

Reviewer 2: ```suggestion
In order to achieve introducing an exact printer, the first step is to preserve concrete syntax information. For example,
```

I don't think you can achieve an object (you don't achieve a house), but rather a goal or an aim. You would use a more concrete action verb instead, and you would achieve that action (What do I want to achieve in 2026? Building a house.)

> comments must not be altered, comma style in comma-separated fields must be preserved, blankspace

Reviewer 2: ```suggestion
comments must not be altered, comma style in comma-separated fields must be preserved, blank spaces
```

and blank lines should be preserved.

[Jappie](https://jappie.me/) started doing a prototype that works well since 2024 Zurihac.
The proposal has matured enough and was accepted by the Haskell Foundation, the current step is

> to implement it in Cabal =D

Reviewer 2: ```suggestion
to implement it in Cabal. =D
```

The following sections will be talking about handling comments in the parser and lexer. The current

> parser doesn't handle comments at all. In fact, the lexer drops them and the parser never sees them.

Reviewer 2: The sentence "The current parser doesn't handle comments at all. In fact, the lexer drops them and the parser never sees them." is happening too quickly, because you only define the terms "parser" and "lexer" _after_. It summarises the situation only if you already know about it. It would be more appropriate to put it further down.

That being said, you reiterate the point after with "Prior to the work on comment-preserving parsing, comments are dropped by the lexer and no other parts down the pipeline can see any comments.", so maybe it is redundant. :)

Author: I think it's indeed redundant and removed it, nice catch!

This inner format is handled by the field grammar parser.
It is a lot less straightforward than the two aforementioned parts and is primarily defined with
type classes. The `FieldGrammar` type class describes how the pretty printer and parser should
behave, two birds one stone! [^2]

# How we achieved comment-preserving parsing
Prior to the work on comment-preserving parsing,
comments are dropped by the lexer and no other parts down the pipeline can see any comments.
To preserve comments, we must go against this prior design by making the lexer emit comments.

We also need to store the comments somewhere in the output of `readFields` which outputs
`[Field Position]`.

The `Field ann` data type is defined as follows
```hs
data Field ann
  = Field !(Name ann) [FieldLine ann]
  | Section !(Name ann) [SectionArg ann] [Field ann]
  deriving (Eq, Show, Functor, Foldable, Traversable, Generic)
```

Originally, Jappie's prototype adds another constructor `Comment !ByteString` to hold a
comment. This was quite straightforward (exactly what we want in a prototype!) but it also
introduced subtle footguns: all code that depend on the `Field` data type now has to deal with
with `Comment`s.

> With help from Andrea, we changed the design to maintain the original definition of `Field ann`,

Reviewer 2: > With help from Andrea

Andrea Leopardi? Andrea Bocelli? Andrea Vezzosi? Andrea Bedini?

while instantiating `ann` differently.
Previously this type variable was `Position`; by changing it to `([ByteString], Position)`, we were
able to store comments without polluting all usages of `Field ann` with the `Comment` constructor!
Concretely we decided to give this pattern a type to disallow uses of `Data.Bifunctor.first` and
`Data.Bifunctor.second`, as it can be less straightforward.

> The `WithComments ann` type is as follows

Reviewer 2: ```suggestion
The `WithComments ann` type is as follows:
```

```hs
data WithComments ann = WithComments
  { justComments :: ![Comment ann]
  , unComments :: !ann
  }
  deriving (Show, Generic, Eq, Ord, Functor)
```

> A minor downside with comments-in-`ann` model is that a file with no field but only comments

Reviewer 2: ```suggestion
A minor downside with the _comments-in-`ann`_ model is that a file with no field but only comments
```

can't be represented. However we don't consider this case because such a file won't be a valid cabal
file.

> We change the top-level field parser `readFields` as follows:

Reviewer 2: ```suggestion
We changed the top-level field parser `readFields` as follows:
```

The jump between preterite and present is inconsistent, I would advise to stick to one tense throughout the file. :)

```hs

> -- Old binding is kept with the same type

Reviewer 2: ```suggestion
-- Old definition is kept with the same type
```

(here `binding` refers to a detail that is too precise in this context to be useful, you want to show how definitions are changed, not the binding type — let vs. where, for instance).

Otherwise, a `GenericPackageDescription` would be considered different when the comments are
different!

We duplicate the top-level parsing functions so we have one entry-point that parses
annotation (`parseAnnotatedGenericPackagedescription`), and another one that doesn't
(`parseGenericPackageDescription`) to provide backward compatibilty.

```hs
data AnnotatedGenericPackageDescription = AnnotatedGenericPackageDescription
  { exactComments :: Map Position ByteString
  , unannotatedGpd :: GenericPackageDescription
  }
  deriving (Show, Data, Generic)
```

Since the position of all comments should not overlap, we create a map out of the comments indexed
by their `Position` in the source file. This is necessary to reconstruct the comments in the source
file and achieve exact print.

This change to the lexer and parser is isolated, meaning you can already use the comments when this is
merged. However more work needs to be done for a complete exact print feature!
Namely we need to stop common stanzas from being merged in.
Currently I am working on a prototype to retain trivia in field grammar parser, so the spaces in
dependency bounds won't be dropped, for example.

> Please tell me if you have any thoughts on this work :)

Reviewer 2: ```suggestion
Please tell me if you have any thoughts on this work. :)
```

I want to thank the the Haskell Foundation and Jappie for allowing this work to happen.
I appreciate greatly the opportunity to work on cabal and solve a problem that helps Haskell

> programmers :)

Reviewer 2: ```suggestion
programmers. :)
```

I also want to thank the people in my life who believed in me when I needed help, without you I

> wouldn't be who I am today :)

Reviewer 2: ```suggestion
wouldn't be who I am today. :)
```
